### PR TITLE
Machine-readable shared wallet API errors.

### PR DESCRIPTION
**Pull request by jonathanknowles**
## Issue Number

Follow-on from review of https://github.com/input-output-hk/cardano-wallet/pull/3676.

## Summary

This PR removes an **interpolated string equality check** from our API integration test suite, and replaces it an equality check that uses **`decodeErrorInfo`** in conjunction with an **API error record**:

```patch
lib/wallet/integration/src/Test/Integration/Scenario/API/Shared/Wallets.hs
 
-     expectErrorMessage (errMsg403NoSuchCosigner (toText Payment) 7) rPatch5
+     decodeErrorInfo rPatch5 `shouldBe` SharedWalletNoSuchCosigner
+         ApiErrorSharedWalletNoSuchCosigner
+             { cosignerIndex = ApiCosignerIndex 7
+             , credentialType = ApiCredentialType Payment
+             }
```
```patch
lib/wallet/integration/src/Test/Integration/Framework/TestData.hs
 
- errMsg403NoSuchCosigner :: Text -> Int -> String
- errMsg403NoSuchCosigner cred cosigner = unwords
-     [ "It looks like you've tried to add a cosigner key to a"
-     , "shared wallet's"
-     , unpack cred
-     , "template for a non-existing cosigner index:"
-     , show cosigner
-     ]
```
The above definition of `errMsg403NoCosigner` (now removed) was essentially a **duplicate** of an identical interpolated string in `Http.Server.Error`:

https://github.com/input-output-hk/cardano-wallet/blob/1263e6ed8c486f31ed75219b59a8e5fc6af577de/lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs#L801-L807

## Advantages

1. If we make any breaking changes to the API error type, the compiler will warn us to update any areas of the test suite that use this type.
2. Someone browsing the integration tests can easily jump to the definition of the error type, or to the definitions of any of the constructors used.
3. We no longer need to manually keep two copies of an interpolated string function in sync with one another.

## Method

To achieve this, we introduce the following structured error type:
```haskell
data ApiErrorSharedWalletNoSuchCosigner = ApiErrorSharedWalletNoSuchCosigner
    { cosignerIndex
        :: !ApiCosignerIndex
    , credentialType
        :: !ApiCredentialType
    }
    deriving (Data, Eq, Generic, Show, Typeable)
    deriving (FromJSON, ToJSON)
        via DefaultRecord ApiErrorSharedWalletNoSuchCosigner
    deriving anyclass NFData
```
And add it as a parameter to the relevant constructor of ApiErrorInfo:
```patch
lib/wallet/api/http/Cardano/Wallet/Api/Types/Error.hs

  data ApiErrorInfo
      = ...
      | ...
-     | SharedWalletNoSuchCosigner
+     | SharedWalletNoSuchCosigner !ApiErrorSharedWalletNoSuchCosigner
      | ...
```
When errors of this type are encoded as JSON, this results in objects of the following format:
```json
{ "code": "shared_wallet_no_such_cosigner"
, "info":
  { "cosigner_index": 182
  , "credential_type": "delegation"
  }
, "message": "It looks like you've..."
}
```

## Future Work

This PR serves as a demonstration of this technique, and only converts one particular error type. We could use this method to remove all duplicate interpolated error messages from our API integration test suite.

### CODE REVIEW

**Review by piotr-iohk — APPROVED** [review]
LGTM

**Review by paweljakubas — APPROVED** [review]
beautiful change!

File: lib/test-utils/src/Test/QuickCheck/Extra.hs [hunk 1]
```diff
@@ -207,6 +208,36 @@ genSized2With f genA genB = uncurry f <$> genSized2 genA genB
...
+--
+shrinkBoundedEnum :: forall a. (Eq a, Enum a, Bounded a) => a -> [a]
+shrinkBoundedEnum a
+    | a == minBound =
+        []
+    | otherwise =
+        toEnum <$> filter (>= minBoundInt) (shrinkIntegral $ fromEnum a)
+  where
+    minBoundInt :: Int
+    minBoundInt = fromEnum (minBound @a)
```
Comment (jonathanknowles) [inline lib/test-utils/src/Test/QuickCheck/Extra.hs:239]: Later on I turned this into a PR on `QuickCheck` itself:
https://github.com/nick8325/quickcheck/pull/350

**Pull request merged**